torchcsprng/csrc/csprng.cpp

```diff
@@ -6,4 +6,373 @@
+      case at::ScalarType::Double:  // Unlikely to happen, but doesn't hurt to check
+        TORCH_CHECK(n <= (int64_t(1) << 53) + 1, "n cannot be greater than 2^53+1 for Double type.");
+        break;
+      default:
+        break;
+    }
+  }
+
+  template <typename scalar_t, typename RNG>
+  void randperm(Tensor& result, int64_t n, c10::optional<at::Generator> generator) {
+    auto gen = at::check_generator<RNG>(generator);
+    scalar_t *r__data = result.data_ptr<scalar_t>();
+
+    result.resize_({n});
+    int64_t r__stride_0 = result.stride(0);
+
+    at::parallel_for(0, n, internal::GRAIN_SIZE,
+                     [&r__data, &r__stride_0](int64_t p_begin, int64_t p_end) {
+                       for(int64_t i = p_begin; i < p_end; i++)
+                         r__data[i*r__stride_0] = static_cast<scalar_t>(i);
+                     });
+
+    for(int64_t i = 0; i < n - 1; i++)
+    {
+      int64_t z = gen->random() % (n-i);
+      scalar_t sav = r__data[i*r__stride_0];
+      r__data[i*r__stride_0] = r__data[(z+i)*r__stride_0];
+      r__data[(z+i)*r__stride_0] = sav;
+    }
+  }
+} // namespace
+
+Tensor& randperm_generator_out(Tensor& result, int64_t n, c10::optional<Generator> generator) {
+  TORCH_CHECK(n >= 0, "n must be non-negative, got", n);
+  check_supported_max_int_with_precision(n, result);
+  if (result.device().type() == at::kCUDA) {
+    auto result_cpu = at::empty({n}, result.options().device(kCPU));
+    randperm_generator_out(result_cpu, n, generator);
+    result.resize_({n});
+    return result.copy_(result_cpu);
+  }
+  result.resize_({n});
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(generator->mutex());
+  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, result.scalar_type(), "randperm", [&]() -> void {
+    randperm<scalar_t, CSPRNGGeneratorImpl>(result, n, generator);
+  });
+  return result;
+}
+
+// ================================================Encrypt/Decrypt=====================================================
+
+Tensor encrypt_pybind(Tensor input, Tensor output, Tensor key, const std::string& cipher, const std::string& mode) {
+  if (input.device().type() == DeviceType::CPU) {
+    return cpu::encrypt(input, output, key, cipher, mode);
+#ifdef WITH_CUDA
+  } else if (input.device().type() == DeviceType::CUDA) {
+    return torch::csprng::cuda::encrypt(input, output, key, cipher, mode);
+#endif
+  } else {
+    TORCH_CHECK(false, TENSOR_DEVICE_TYPE_IS_NOT_SUPPORTED);
+  }
+}
+
+Tensor decrypt_pybind(Tensor input, Tensor output, Tensor key, const std::string& cipher, const std::string& mode) {
+  if (input.device().type() == DeviceType::CPU) {
+    return cpu::decrypt(input, output, key, cipher, mode);
+#ifdef WITH_CUDA
+  } else if (input.device().type() == DeviceType::CUDA) {
+    return torch::csprng::cuda::decrypt(input, output, key, cipher, mode);
+#endif
+  } else {
+    TORCH_CHECK(false, TENSOR_DEVICE_TYPE_IS_NOT_SUPPORTED);
+  }
+}
```

Comment:
are these functions pybind specific, can you call them encrypt/decrypt for consistency with the rest?